### ISSUE TITLE
Root Class Needs Refactoring

### ISSUE THREAD

**Issue by hut104** [APSIMInitiative/ApsimX#5489]
The root class has been under development as part of work on the sorghum model.  As a result, several algorithms have been incorporated into the class code rather than adding new functionality through use of functions (to capture differing options).

Furthermore, an optional switch has been added for calculation of a root front (probably for lateral root growth). However this optional switch is used to switch between all options (ie it is a single switch for options used in the sorghum model as far as I can see).  For example, if the the root front switch is present, the model appears to use a different N uptake approach.

The root model needs to be completely refactored to remove these sorts of developments.  All options should be captured using functions in the Json rather than optional flags within the code.
Finally, all optional properties should be removed or made mandatory.

**Comment by jbrider** [comment]
@hut104 I thought we had got it to a point at the last sprint where it wasn't holding you up - that we had enabled the functions to be extracted out?  Willing to catch up and discuss refactoring either way.  My understanding of multiple root zones is poor - you have mentioned previously that we could do skip row using them - is there an example of a skip row setup we could use to test with?

**Comment by hut104** [comment]
Yep.  We've been able to keep going.  But the code really does need to be cleaned up.  I see the same calculations occurring throughout the code - never a good sign.  There are also flags to allow optional execution of alternate methods etc.  These need to be fixed to follow out guidelines for maintainability.

The strip crop examples are in the microclimate and agroforestry examples (e.g. gliricidia) and tests.  We will probably use these in the cotton model development.

Neil.

________________________________
